SQLite storage optimized by default

adamdbrw opened the pull request:
With optimized settings applied, performance of storage writing is significantly increased. This reduces the problem of messages being lost when there is a high throughput of recorded data.

![image(2)](https://user-images.githubusercontent.com/16702721/100106722-30cb3680-2e69-11eb-92f9-59b18dd346e9.png)

Note that before this PR optimization is already accessible to users through `--storage-config-file settings`. This PR makes storage optimization the default. 

Includes an option to use former settings: `--resilient-storage-writing`

adamdbrw commented:
@mjeronimo and/or @Karsten1987, could you review this one?

mjeronimo commented on ros2bag/ros2bag/verb/record.py:
```diff
@@ -90,6 +90,13 @@ def add_arguments(self, parser, cli_name):  # noqa: D102
         )
+        parser.add_argument(
+            '--resilient-storage-writing', action='store_true',
+            help='Indicate preference for avoiding data corruption in case of crashes, '
```
Could clarify here that --resilient-storage-writing overrides only corresponding flags; additional flags from the config file remain in the pragmas list. I believe this is what is meant by "can override," but perhaps we can make that sentence a bit more explicit. For example, "When set, this flag will override corresponding settings in the config passed with --storage-config-file."

mjeronimo replied:
Related, I'd like to see a test for the above behavior to ensure that overriding of params in a config file by using --resilient-storage-writing is occurring as expected.

adamdbrw replied:
Applied - take a look at the unit test. I preferred to extend the sqlite storage API to enable querying of current settings (which seems useful anyway) because testing otherwise would be awkward.

mjeronimo commented on rosbag2_storage_default_plugins/test/rosbag2_storage_default_plugins/sqlite/test_sqlite_storage.cpp:
```diff
@@ -322,6 +322,28 @@ TEST_F(StorageTestFixture, loads_config_file) {
+  // resilient_storage_writing should replace "some_value" with "wal
+  EXPECT_EQ(writable_storage->get_storage_setting("journal_mode"), "wal");
+
+  // resilient_storage_writing should replace "another_value" with "2"
```
Comment does not match code. The comment mentions setting "another_value" to "2", while the code is checking against "1".

adamdbrw replied:
The correct value for default is NORMAL (returned as 1), so the comment is mistaken (commented before I checked) and the code is right.

Fixed

https://www.sqlite.org/pragma.html#pragma_synchronous

mjeronimo commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=13181)](http://ci.ros2.org/job/ci_linux/13181/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=8112)](http://ci.ros2.org/job/ci_linux-aarch64/8112/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10897)](http://ci.ros2.org/job/ci_osx/10897/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=13195)](http://ci.ros2.org/job/ci_windows/13195/)

emersonknapp reviewed (changes requested):
I'm back from long holiday weekend - sorry for the slow response. My main comment is the structural concern of making the storage settings profiles more extensible.

adamdbrw commented on ros2bag/ros2bag/verb/record.py:
```diff
@@ -90,6 +90,13 @@ def add_arguments(self, parser, cli_name):  # noqa: D102
             help='Path to a yaml file defining overrides of the QoS profile for specific topics.'
         )
+        parser.add_argument(
+            '--resilient-storage-writing', action='store_true',
```
@emersonknapp that would likely be better in the long run, if there are sensible presets other than optimized/safer. Is this a blocker for this PR? The primary goal of this PR is to change the default optimization for the storage.

adamdbrw replied:
@emersonknapp I am fine with either, please let me know if I should put it into this PR or a separate one.

emersonknapp replied:
I would prefer to do this change now - the `resiliant_storage_writing` value is the sort of thing that seems likely to linger even though it is an overly narrow API. Calling is a `preset_config` probably also does a good job of clarifying that the `storage-config-file` overrides these values.

If you will do the followup change, though, I won't block this PR.

adamdbrw replied:
@emersonknapp I made an attempt to implement it, but there are some drawbacks to each solution I came up with. There is currently no ros2_storage interface exposed to the python (except for some structs) and, since this API requires plugin-specific return value, one would need to instantiate storage to return these profiles. Which would be the only case when this is done, but the storage remains not-open and is discarded after the call. 

Another solution would be to have a separate object hierarchy (and a factory) just for the profiles.

Yet another would be to simply keep a dictionary of plugin name - preset profiles at the top level, and return based on which plugin is meant to be used (as in the argument).

Each solution has drawbacks and I would love to discuss that with you. Did you have a particular design in mind? We could also set up a call if that would make things easier, let me know.

emersonknapp commented on README.md:
```diff
@@ -125,8 +125,13 @@ read:
+
+By default, SQLite settings are significantly optimized for performance.
+This might have consequences of bag data being corrupted after an application or system-level crash.
+If increased crash-caused corruption resistance is necessary, use `--resilient-storage-writing` flag.
```
Should note that `any values specified in the `--storage-config-file` take precedence over default or preset values.

emersonknapp commented on README.md:
```diff
@@ -125,8 +125,13 @@ read:
-Please refer to [documentation of pragmas](https://www.sqlite.org/pragma.html).
+
+By default, SQLite settings are significantly optimized for performance.
+This might have consequences of bag data being corrupted after an application or system-level crash.
```
It would be good to clarify that only the current bagfile has the possibility to be corrupted in this case, any already-split files are safe once finalized, so the maximum data loss is bounded by the maximum bag (time or size)

emersonknapp commented on ros2bag/ros2bag/verb/record.py:
```diff
@@ -90,6 +90,13 @@ def add_arguments(self, parser, cli_name):  # noqa: D102
+            help='Indicate preference for avoiding data corruption in case of crashes, '
+                 'at the cost of performance. Setting this flag disables optimization '
+                 'settings for storage (the defaut). When set, this flag will override '
+                 'corresponding settings in the config passed with --storage-config-file.'
```
I think that this should not have top precedence, logically. It should modify the baseline settings, anything the user passes _explicitly_ (the `storage-config-file`) should always be respected.

adamdbrw commented:
Fixing the merge - went a bit sideways

emersonknapp commented:
Running for evidence - we can override my comments for now if @mjeronimo you'd rather move forward this way and have us address that stuff later

Gist: https://gist.githubusercontent.com/emersonknapp/c32bfc13fc1341e9bdaa088b4ef23bf1/raw/34a8fb38e8af97a7e3f68337d715b0a0c1411b52/ros2.repos
BUILD args: --packages-up-to ros2bag rosbag2_storage rosbag2_storage_default_plugins rosbag2_transport
TEST args: --packages-select ros2bag rosbag2_storage rosbag2_storage_default_plugins rosbag2_transport
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/7389
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=13345)](http://ci.ros2.org/job/ci_linux/13345/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=8273)](http://ci.ros2.org/job/ci_linux-aarch64/8273/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=11075)](http://ci.ros2.org/job/ci_osx/11075/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=13397)](http://ci.ros2.org/job/ci_windows/13397/)

adamdbrw commented:
@emersonknapp I addressed the API issue - should remain the same now and we are free to add new presets (or change the way we acquire presets from a simple static list as it is currently). 

Also, config file is now overriding the preset as you suggested. I also added explanation that the crash-caused corruption risk only applies to current bagfile if splitting is on.

Looking into Windows issues now.

emersonknapp reviewed (approved):
LGTM! Thanks for the changes. Running CI now

emersonknapp commented:
Gist: https://gist.githubusercontent.com/emersonknapp/b8abd49ac4aa32d70feee3fa3b3b06b3/raw/34a8fb38e8af97a7e3f68337d715b0a0c1411b52/ros2.repos
BUILD args: --packages-up-to ros2bag rosbag2_storage rosbag2_storage_default_plugins rosbag2_transport rosbag2_tests
TEST args: --packages-select ros2bag rosbag2_storage rosbag2_storage_default_plugins rosbag2_transport rosbag2_tests
Job: ci_launcherci_launcher ran: https://ci.ros2.org/job/ci_launcher/7421
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=13383)](http://ci.ros2.org/job/ci_linux/13383/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=8305)](http://ci.ros2.org/job/ci_linux-aarch64/8305/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=11108)](http://ci.ros2.org/job/ci_osx/11108/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=13433)](http://ci.ros2.org/job/ci_windows/13433/)
